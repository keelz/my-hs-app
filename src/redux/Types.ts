import { MiddlewareAPI, Dispatch } from 'redux';
import { IBrowser } from 'redux-responsive';
import { AppState } from './reducers/App';
import { ICollectionState } from './reducers/Collection';
import { ICardsState } from './reducers/Cards';

export interface IActionType<T> {
  type: string;
  payload?: T;
}

export interface IRootState {
  App: AppState;
  Browser: IBrowser;
  Cards: ICardsState;
  Collection: ICollectionState;
}

export type AppMiddlewareApi = MiddlewareAPI<Dispatch<IActionType<any>>, IRootState>;

export type AsyncMiddlewareOperation = (api: AppMiddlewareApi) =>
  (next: Dispatch<IActionType<any>>) =>
  (action: IActionType<any>) => Promise<any>;

export type MiddlewareOperation = (api: AppMiddlewareApi) =>
  (next: Dispatch<IActionType<any>>) =>
  (action: IActionType<any>) => void;

export type SafeMiddlewareOperation = (api: AppMiddlewareApi) =>
  (action: IActionType<any>) => void;

export type CollectionFilters = {
  [field: string]: string | string[]
};

export enum MODAL_STATE { CLOSED, OPEN }

export enum PLAY_STYLE {
  STANDARD = 'STANDARD',
  WILD = 'WILD',
}
