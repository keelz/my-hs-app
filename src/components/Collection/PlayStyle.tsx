import * as React from 'react';
import { IComponentProps } from '../../common/models/App.model';
import { composeClassname } from '../../common/utils';

const wildIcon = require('../../common/assets/img/99px-wild_icon_large.png');
const standardIcon = require('../../common/assets/img/100px-standard_icon_large.png');

interface IPlayStyleProps extends IComponentProps {}

interface IPlayStyleStateProps {
  filters: { [field: string]: string };
  setFilter: (name: string, value: string) => any;
}

type Props = IPlayStyleProps & IPlayStyleStateProps;

const PlayStyle: React.SFC<Props> = props =>
  <div className={composeClassname([
    'Collection-play-style',
  ])(props.className)}>
    { props.filters['PLAY_STYLE'] === 'WILD'
      ? <img
        src={wildIcon}
        alt="wild-ico"
        onClick={() => props.setFilter('PLAY_STYLE', 'STANDARD')}
        title="Filter Standard Cards" />
      : <img
        src={standardIcon}
        alt="standard-ico"
        onClick={() => props.setFilter('PLAY_STYLE', 'WILD')}
        title="Filter Wild Cards" />
    }
  </div>;

export default PlayStyle;
