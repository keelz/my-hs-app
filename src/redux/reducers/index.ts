import { combineReducers } from 'redux';
import { responsiveStateReducer as Browser } from 'redux-responsive';
import appReducer from './App';

const rootReducer = combineReducers({
  app: appReducer,
  browser: Browser,
});

export default rootReducer;
