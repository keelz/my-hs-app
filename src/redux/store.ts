import { createBrowserHistory } from 'history';
import { routerMiddleware } from 'react-router-redux';
import { applyMiddleware, createStore, compose } from 'redux';
import { responsiveStoreEnhancer } from 'redux-responsive';
import createDebounce from 'redux-debounced';
import thunk from 'redux-thunk';
import appMiddleware from './middleware/App';
import rootReducer from './reducers';

/**
 * redux devtools implementation.
 * We want to compose our enhancers normally within the production
 * environment (build), otherwise we want to compose our ehancers
 * with the devtools extension.
 */
export const composeFunc = () => {
  return process.env.NODE_ENV === 'production'
    ? compose
    : (() => {
      const { composeWithDevTools } = require('redux-devtools-extension');
      return composeWithDevTools;
    })();
};

/**
 * Third party redux middleware
 * - redux-thunk
 * - router
 * - react-router-redux
 * - history
 * - redux-responsive
 */
const libEnhancers = [
  thunk,
  routerMiddleware(createBrowserHistory({ basename: '/' })),
  createDebounce(),
].map(mw => applyMiddleware(mw));
libEnhancers.push(responsiveStoreEnhancer);

/**
 * Custom middleware
 * - appMiddleware
 */
const customEnhancers = [
  appMiddleware,
].map(mw => applyMiddleware(mw));

// compose enhancers with composeFunc
// https://redux.js.org/api/compose
const composedEnhancers = composeFunc().apply(null, [...libEnhancers, ...customEnhancers]);

// create store
// https://redux.js.org/api/createstore
const store = createStore(rootReducer, {}, composedEnhancers);

export default store;
