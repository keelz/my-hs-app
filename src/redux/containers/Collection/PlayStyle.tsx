import { connect } from 'react-redux';
import PlayStyle from '../../../components/Collection/PlayStyle';
import { IRootState } from '../../Types';
import { collectionSetFilterAction } from '../../reducers/Collection';

const mapStateToProps = (state: IRootState) => ({
  filters: state.Collection.filters,
});

const mapDispatchToProps = {
  setFilter: collectionSetFilterAction,
};

const ConnectedPlayStyle = connect(mapStateToProps, mapDispatchToProps)(PlayStyle);

export default ConnectedPlayStyle;
