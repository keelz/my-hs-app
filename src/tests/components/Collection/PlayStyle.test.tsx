import * as React from 'react';
import * as enzyme from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import APP from '../../../common/constants/app';
import PlayStyle from '../../../components/Collection/PlayStyle';
import { PLAY_STYLE } from '../../../redux/Types';

const { FILTERS } = APP.COLLECTION;

const defaultProps = Object.freeze({
  filters: {
    [FILTERS.PLAY_STYLE]: PLAY_STYLE.STANDARD,
  },
  setFilter: jest.fn(),
});

describe('Collection/PlayStyle', () => {
  describe('snapshots', () => {
    it('renders without crashing', () => {
      const wrapper = enzyme.shallow(<PlayStyle {...defaultProps} />);
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('integration', () => {
    it(`handles onClick event correctly with ${PLAY_STYLE.STANDARD} play style`, () => {
      const props = {
        ...defaultProps,
        setFilter: jest.fn(),
      };
      const wrapper = enzyme.mount(<PlayStyle {...props} />);
      const img = wrapper.find('img');
      img.simulate('click');
      expect(props.setFilter).toHaveBeenCalledWith(
        FILTERS.PLAY_STYLE,
        PLAY_STYLE.WILD
      );
    });

    it(`handles onClick event correctly with ${PLAY_STYLE.WILD} play style`, () => {
      const props = {
        ...defaultProps,
        filters: {
          ...defaultProps.filters,
          [FILTERS.PLAY_STYLE]: PLAY_STYLE.WILD,
        },
        setFilter: jest.fn(),
      };
      const wrapper = enzyme.mount(<PlayStyle {...props} />);
      const img = wrapper.find('img');
      img.simulate('click');
      expect(props.setFilter).toHaveBeenCalledWith(
        FILTERS.PLAY_STYLE,
        PLAY_STYLE.STANDARD
      );
    });
  });
});
