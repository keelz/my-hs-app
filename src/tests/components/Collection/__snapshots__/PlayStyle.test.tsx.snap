// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection/PlayStyle snapshots renders without crashing 1`] = `
<div
  className="Collection-play-style"
>
  <img
    alt="standard-ico"
    onClick={[Function]}
    src="100px-standard_icon_large.png"
    title="Filter Wild Cards"
  />
</div>
`;
